### ISSUE TITLE
Linked raster images in vector layers

### ISSUE THREAD

**Issue by snan** [mypaint/mypaint#519]
OK, I've tested this out now:
Mypaint does show embedded raster images without problem (great!) but it doesn't show linked images (even in vector layers it "created" and then inkscape added the images).

Weirdly enough, unzipping the ora and opening the svg in `data/` with inkview shows both images. So my first hunch, that the pathname had been mangled, was wrong.

As an aside, inkscape allows editing the raster images with mypaint so there's a little bit of матрёшка doll going on, kinda cool and can be useful for making masks, light maps etc. Both programs can spawn each other.

**Comment by achadwick** [comment]
Yeah, probably a limitation of librsvg. Can the `rsvg` command-line tool manage these files?

Inkscape's SVG is a little unconventional. Does saving the file as "Plain SVG" change the way files with external linked parts work? /hoping it inlines all the things/

I don't think SVG-based vector layers should ever contain externally linked resources. Although I confess I don't know what the consequences of that really are for things like fonts.

Bottom line here is that OpenRaster doesn't have a resolving scheme for embedded clever documents. After all, the document may never actually be unpacked to somewhere other than RAM.

**Comment by snan** [comment]
Yeah, I saved it as plain svg without sodipodi or inkscape stuff.
The `rsvg` command-line tool has the same behaviour as Mypaint, i.e. it shows embedded but not linked.
So does `eog`.

The web browsers I have installed shows both, and (of course) inkview and inkscape.

Guess this is another "upstream".

As for fonts, this might mean that you rely on the user having fonts installed with matching names and that you can't use webfonts.

**Comment by snan** [comment]
I've found great use for having raster images in the SVGs; you can do some quick 5 minute sketches in mypaint for each comic panel,
then lay them out in Inkscape (i.e. every panel a separate image file, you can move, rotate, crop etc), and letter, draw frames, word balloons etc in Inkscape, then bring it back to Mypaint (with the vector layers still being vector) to ink and color it.
Sometimes you need to transform texts to paths though :/ if the font is not one that rsvg knows about.
Summary: all the problems are upstream,
Temporary workaround: embed images, turn texts to paths.

**Comment by achadwick** [comment]
Just perusing the docs for now. rsvg_handle_set_base_uri() or rsvg_handle_set_base_gfile() may be the functions to use here, assuming we can get an RsvgHandle* out of GdkPixbuf's internals.

rsvg_set_default_dpi() may be of use for font scaling issues. There's an RsvgHandle variant of it too.

/infodump

**Comment by snan** [comment]
Just wanna quickly add (but maybe I already said that) that I've realized that the fact that mypaint uses rasterized versions of the svg layers is an amazing feature, not a bug (esp since you can still edit them). I can put in so much more than I could back when I was only using inkscape, it would've died under the weight.

**Comment by jplloyd** [comment]
librsvg does support linking, and has since 2013, but for security reasons it restricts linked items to things in the same directory, or directories below the location of the main document. 

My understanding of the vector layer implementation is that the svg data is stored in a temporary folder (when extracted), so no linked data could ever fulfil the subtree requirement without creating hard links using extracted paths, or similar ugly solutions.

So it's not so much an upstream issue, as it is an inherent limitation from the choice of library (otherwise a very nice and convenient choice) and the limitations of the OpenRaster spec (if permitted by the spec, an alternative would be to have a layer type that just reads an existing svg from the disk, only storing the path in the layer data - link subtree limitations would still apply with librsvg of course).
